q2_metadata/normalization/_norm_check_rules.py

```diff
@@ -0,0 +1,311 @@
+
+from q2_metadata.normalization._norm_errors import (
+    RuleError, ExpectedError, OntologyError, RemapError,
+    ValidationError, NormalizationError,
+    BlankError, MissingError, FormatError
+)
+
+
+def check_rule(variable: str, rules: dict, rule: str, rule_value):
+    """For each rule in the current variable's rules set,
+    perform checks that the rule values to be used for the rules
+    to apply are properly formatted.
+        - If properly formatted: full the rules dict
+        for the variable's Rules() class instance.
+        - If not properly formatted: raise a specific error and
+        collect the encountered error in the ErrorsCollection()
+        instance for this variable's Rules() class instance.
+
+    Parameters
+    ----------
+    variable : str
+        Name of the metadata variable for which there is rules.
+    rules : dict
+        Rules in the hard-coded data structure that has
+        common default values for all variables and
+        that is checked and updated here.
+    rule : str
+        Current rule name. Could be one of these:
+        - expected: accepted, controlled vocabulary from AGP.
+        - ontology: accepted, controlled vocabulary from a third-party resource.
+        - remap: the replacements to perform ( "A" -> "B" )
+        - validation: the
+        - normalization:
+        - blank:
+        - missing:
+        - format:
+    rule_value : str, list or dict
+        Parsed rule value for the current rule.
+    """
+    rule_type_map = {
+        'expected': 'lookups',
+        'ontology': 'lookups',
+        'remap': 'edits',
+        'validation': 'edits',
+        'normalization': 'edits',
+        'blank': 'allowed',
+        'missing': 'allowed',
+        'format': 'allowed'
+    }
+
+    if rule not in rule_type_map:
+        raise RuleError(rule, variable, rule_value)
+
+    if rule in ['blank', 'missing']:
+        rule_method_key = 'check_blank_missing'
+    else:
+        rule_method_key = 'check_%s' % rule
+
+    rule_method = globals()[rule_method_key]
+    rule_error = globals()['%sError' % rule.capitalize()]
+    rule_type = rule_type_map[rule]
+    if rule_method(rule_value):
+        rules[rule_type][rule] = rule_value
+    else:
+        raise rule_error(variable, rule_value)
+
+
+def check_expected(rule_value):
+    """Check that the user-defined value for the
+    rule "expected" is correctly formatted.
+
+    expected must be a list
+        Apply rule of allowing only expected values:
+        These do allow the missing and blanks.
+                    expected:
+                        - str
+                        - list
+
+    Parameters
+    ----------
+    rule_value
+        Parsed rule value for the rule "expected".
+
+    Returns
+    -------
+    bool
+        True if successful, False otherwise.
+    """
+    # check if the rule value is indeed a list (it must be the case).
+    if isinstance(rule_value, list):
```

Comment:
If type checking is required, it suggests the data aren't being properly sanitized higher in the stack

Reply:
This is the check... I have to check that the format passed by the user is correct and I have not idea how to do that otherwise. It's been review by @ElDeveloper and I cannot sanitized a human ;)